# Supporting software

## Paper

* Title: A protein probability model for high-throughput protein identification by mass spectrometry-based proteomics
* Authors: Gorka Prieto and Jesús Vázquez
* Contact: Gorka Prieto <gorka.prieto@ehu.eus>
* Journal: Journal of Proteome Research, 2019

## Source code and binaries

The executable code is available at:
* https://github.com/akrogp/SpHPP/raw/master/dist/LPGF/lpgf-v1.0.1.tar.gz

The source code is available at:
* EhuBio: https://github.com/akrogp/EhuBio
* SpHPP: https://github.com/akrogp/SpHPP

## System requirements

* UNIX shell (bash): for running the principal scripts (.sh) created for convenience
* Make (GNU make): for using the workflow configuration provided (workflow.cfg)
* Java 8: for executing core processing using the provided EhuBio.jar file
* Python 3: for executing optional fasta database utilities

## Fasta dababase

Separated target and decoy database searches are required by this software. The target-decoy competition is carried out *a posteriori* by the software itself.

Protein accessions and gene names are required to be present in the target fasta file headers. UniProt headers format is recommended for that.

Decoy headers must contain only the decoy name using a prefix prepended to their corresponding target protein accessions.

A `db.sh` script is provided for easily generating a target and a decoy database using the same criteria we used in the paper. In the target database isoleucines are replaced with leucines, and the decoy database is generated using `decoyPYrat`.

```
Usage:

  ./db.sh input.fasta decoy_prefix

  Where:
    input.fasta : input fasta file with target sequences using UniProt headers
    decoy_prefix: prefix to be used for decoy names

  Outputs:
    target.fasta: output target fasta file replacing isoleucines in input.fasta
    decoy.fasta : output decoy fasta file generated with decoyPYrat
```

If you prefer to avoid this step you can use the example target and decoy fasta files available here: https://ehubox.ehu.eus/s/mxtKY7iRPNiGprq?path=%2Finput%2Fdb

## Gene-level scores and FDR

### Executing the tool

A `run.sh` script is provided for easily running all the gene-level scores and FDRs combinations presented in the paper using the fasta target/decoy databases and the search engine output files provided by the user.

It is important to use the target/decoy fasta headers formatting explained in the previous section. You can use the ones generated by `db.sh` or the examples provided.

```
Usage:

  ./run.sh target.fasta decoy.fasta decoy_prefix target_dir decoy_dir output_dir

  Where:
    target.fasta: fasta file with target sequences using UniProt headers
    decoy.fasta : fasta file with decoy sequences which headers consists ONLY
                  in a decoy prefix prepended to the target protein accession
    decoy_prefix: prefix used for decoy names
    target_dir  : directory with search engine output files using a separated
                  target database search (*)
    decoy_dir   : directory with search engine output files using a separated
                  decoy database search (*)
    output_dir  : directory for saving the results after executing this workflow

    (*) Supported file formats: X!Tandem XML (*.t.xml), pepXML, Sequest MSF,
        Mascot DAT, mzIdentML (*.mzid)
```

If you want to run the tool using sample X!Tandem output files you can use the ones in the `target` and `decoy` directories available here: https://ehubox.ehu.eus/s/mxtKY7iRPNiGprq?path=%2Finput%2Ftandem

### Tool results

A separated chart is generated for each gene-level score in a PDF file. This chart is similar to Figure 1 of the paper and is useful for checking whether the gene-level scores of the decoy genes follow an uniform distribution.

For each combination of the gene-level score types and FDR techniques presented in the paper a TSV (tab-separated values) file is generated. This file provides for each gene the gene-level score, the local FDR and the q-value.

A number of files corresponding to each processing step are also available within the `all` subdirectory of this output folder.

An output result using the example fasta and X!Tandem files provided is available here: https://ehubox.ehu.eus/s/mxtKY7iRPNiGprq?path=%2Foutput